src/protocols/ab/eip_cip.c

```diff
@@ -1090,14 +1090,17 @@ int build_write_request_connected(ab_tag_p tag, int byte_offset)
     data += write_size;
     tag->offset += write_size;
 
-    /* need to pad data to multiple of 16-bits */
+    /* need to pad data to multiple of either 1, 2 or 4 bytes */
     /* for some PLCs (OmronNJ), padding causes issues when writing counted strings as it creates a mismatch between
         the length of the string and the count integer, therefor this padding can be disabled using the str_pad_16_bits attribute */
-    pad_to_even_bytes = tag->byte_order->str_pad_to_16_bits;
-    if (pad_to_even_bytes == 1){
-        if (write_size & 0x01) {
-            *data = 0;
-            data++;
+    str_pad_to_multiple_bytes = (int)tag->byte_order->str_pad_to_multiple_bytes;
+    if ((str_pad_to_multiple_bytes == 2 || str_pad_to_multiple_bytes == 4) && write_size != 0){
+        if (write_size % str_pad_to_multiple_bytes !=0){
+            int pad_size = str_pad_to_multiple_bytes-(write_size % str_pad_to_multiple_bytes);
+            for (int i =0; i<pad_size; i++){
```

Comment:
Should check that we have buffer space.

Reply:
This does seem like something that should be happening, but Im not sure how to check the size here and dont see it being checked anywhere else in this function? Is it req->request_capacity im looking for?

Reply:
Hmm, I am going to have to look at the surrounding code.   The variable "write_size" does hold the amount we are going to write, but that could be constrained by things like a full request buffer.

Also I just realized that I am not sure that this padding is applicable only to strings.   The original code did not care and just padded everything.  That's a problem.   Phwew if this needs to be type-aware... yuck.

Reply:
I have been thinking about this a bit, originally I did only intend this to affect writing strings but it will affect all writes. I have been testing with these changes and not noticed any problems with writing other datatypes on OmronNJ. It seems that at least for Omron, you do not need to write the correct number of padding 0's for the alignment of the datatype. If this was necessary then you would need to add different numbers of 0's for different datatypes (which have different padding rules) and previously the code always padded to an even number of bytes. The only time there is an issue is when we do add padding (for writing strings).

Do different dtypes have different alignments in other PLCs or is this just an Omron thing? Is the idea of this padding to match the alignment of the data or is it just intended to give an even number of bytes? If we removed the padding altogether would writes fail for other PLC types?

If the user specifies this new str_pad_to_multiple_bytes for a non string tag when the documentation will clearly state that it is only for strings, is that a problem we should be worried about? It might not be a problem that this code affects non string dtypes, as the tag which has str_pad_to_multiple_bytes enabled should only be reading a string.